Common/src/main/java/com/blamejared/crafttweaker/api/tag/CraftTweakerTagRegistry.java

```diff
@@ -0,0 +1,113 @@
+package com.blamejared.crafttweaker.api.tag;
+
+import com.blamejared.crafttweaker.api.CraftTweakerAPI;
+import com.blamejared.crafttweaker.api.annotation.ZenRegister;
+import com.blamejared.crafttweaker.api.tag.manager.ITagManager;
+import com.blamejared.crafttweaker.api.tag.manager.type.KnownTagManager;
+import com.blamejared.crafttweaker.api.util.InstantiationUtil;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.TagManager;
+import org.openzen.zencode.java.ZenCodeGlobals;
+import org.openzen.zencode.java.ZenCodeType;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+@ZenRegister
+@ZenCodeType.Name("crafttweaker.api.tag.TagManager")
+public final class CraftTweakerTagRegistry {
+    
+    public static final String GLOBAL_NAME = "tags";
+    
+    @ZenCodeGlobals.Global(GLOBAL_NAME)
+    public static final CraftTweakerTagRegistry INSTANCE = new CraftTweakerTagRegistry();
+    
+    private CraftTweakerTagRegistry() {}
+    
+    private final Map<ResourceKey<? extends Registry<?>>, ITagManager<?>> registeredManagers = new HashMap<>();
+    private final Set<ResourceKey<? extends Registry<?>>> customManagers = new HashSet<>();
+    
+    public <T> ITagManager<T> addManager(Class<? extends ITagManager<T>> cls) {
+        
+        ITagManager<T> manager = InstantiationUtil.getOrCreateInstance(cls);
+        Objects.requireNonNull(manager);
```

Comment:
It might be better to add a descriptive error message here
```java
Objects.requireNonNull(manager, "Unable to create a manager from class " + cls.getName());
```